src/watershed.jl

```diff
@@ -0,0 +1,89 @@
+import Base.isless
+
+immutable PixelKey{CT}
+    val::CT
+    time_step::Int
+end
+isless{T}(a::PixelKey{T}, b::PixelKey{T}) = (a.val < b.val) || (a.val == b.val && a.time_step < b.time_step)
+
+"""
+```
+segments                = watershed(img, [markers])
+``` 
+Segments the image using watershed transform. Each basin formed by watershed transform corresponds to a segment.
+If you are using image local minimas as markers, consider using hmin_transform to avoid oversegmentation.
+    
+Parameters:  
+-    img            = input grayscale image
+-    markers        = an array marking the pixels where flooding should start
+
+"""
+function watershed{T<:Images.NumberLike, S<:Integer, N}(img::AbstractArray{T, N}, markers::AbstractArray{S,N})
+
+    if indices(img) != indices(markers)
+        error("image size doesn't match marker image size")
+    end
+
+    segments = copy(markers)
+    pq = PriorityQueue(CartesianIndex{N}, PixelKey{T}, Base.Order.Forward)
+    time_step = 0
+
+    R = CartesianRange(indices(img))
+    Istart, Iend = first(R), last(R)
+    for i in CartesianRange(size(img))
+        if markers[i] > 0
+            for j in CartesianRange(max(Istart,i-1), min(i+1,Iend))
```

Comment:
Same comment here as in https://github.com/JuliaImages/ImageSegmentation.jl/pull/7#discussion_r127699625. Relevant below too.